### PR TITLE
Show server URL & QR

### LINKED ISSUE #2479: Display server URL and possibly QR code
## Is your feature request related to a problem? Please describe 👀

For ease of configuring the cold chain app, it would be useful if omSupply displayed the server URL somewhere.
Even better would be to display a QR which the CCA could scan and configure itself

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #2479

# 👩🏻‍💻 What does this PR do? 
 
Displays the current server URL in text and QR format on the sync page, to allow for simpler setup of the cold chain app sync:

**Browser**

Screenshot 2023-11-15 at 10 31 45 AM

**Android**
Screenshot 2023-11-15 at 11 58 42 AM

**Desktop**
Screenshot 2023-11-15 at 12 03 43 PM

I also tried using the [qrcode.react](https://github.com/zpao/qrcode.react) package so that I could embed the mSupply guy as an image in the QR code, but the QR (even without the image) was being read less reliably, so have used the [react-qr-code](https://www.npmjs.com/package/react-qr-code) package

I've had to tweak the server discovery for android so that it returns an IP which isn't the loopback IP, as per:

Screenshot 2023-11-15 at 11 57 07 AM

Screenshot 2023-11-16 at 9 54 58 AM

Screenshot 2023-11-16 at 9 55 04 AM

# 🧪 How has/should this change been tested? 

Have run the site in various modes:
- website (develop)
- website (hosted by server)
- android (client)
- android (server)
- desktop (mac, client)

Have checked that the QR code can be read using my phone and a samsung SM-T500

## 📃 Documentation

- [ ] Screenshot and description of the sync page changes

### CODE REVIEW

**Review by andreievg — COMMENTED** [review]
Thanks Mark
I tested with electron, works like a charm.

Tested on mobile, but the server was loopback, is there any extra permissions that we need to add to manifest or ? ([this stack oveflow](https://stackoverflow.com/questions/6064510/how-to-get-ip-address-of-the-device-from-code) seems to suggest as such)

Had a hard time scanning QR code from tablet on my phone. Should we make it click and expand ?

Also was wondering if this really needs to be in 1.5.0 ?

File: client/packages/host/src/components/Sync/ServerInfo.tsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  Box,
+  Typography,
+  frontEndHostUrl,
+  useNativeClient,
+  useTranslation,
+} from '@openmsupply-client/common';
+import QRCode from 'react-qr-code';
+
+const ServerInfoComponent = () => {
+  console.info('app bar??');
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply client/packages/host/src/components/Sync/ServerInfo.tsx:13]: gah! thanks 🙏

File: client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java [hunk 2]
```diff
@@ -344,15 +347,34 @@ public void onResolveFailed(NsdServiceInfo serviceInfo, int errorCode) {
         });
     }
 
+    // Attempt to get a non-loopback address for the local server
+    // and fallback to loopback if there is an error
+    private String getLocalAddress(NsdServiceInfo serviceInfo){
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java:352]: it wasn't for me - I get `127.0.0.1` on android, but haven't seen it happen on electron

File: client/packages/host/package.json [hunk 3]
```diff
@@ -27,18 +27,19 @@
...
     "@openmsupply-client/inventory": "^0.0.0",
     "@openmsupply-client/invoices": "^0.0.0",
     "@openmsupply-client/requisitions": "^0.0.0",
     "@openmsupply-client/system": "^0.0.0",
-    "@openmsupply-client/coldchain": "^0.0.0",
     "@pmmmwh/react-refresh-webpack-plugin": "^0.5.4",
     "history": "^5.1.0",
     "react": "^18.0.0",
     "react-dom": "^18.0.0",
+    "react-qr-code": "^2.0.12",
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply client/packages/host/package.json:42]: I prefer to restrict to the implementing package for specific things like this, if it's only in one package that is. Was thinking that as soon as we use QR generation elsewhere I'd move it to common. Given that this is the host package though, it doesn't save anything but not putting it in common (both host and common are used by everything). Could go either way with this one

**Comment by mark-prins** [comment]
Ah. yes - I see now.. on later android versions you need 
[NetPermission("getNetworkInformation")](https://developer.android.com/reference/java/net/NetworkInterface#getInetAddresses())

and the 'click to expand'! genius 🙏 

I'm keen to get into 1.5.00 just because it's taken so long to get the release out and I'm concerned that if I don't get this in now, it'll be ages before we can implement the QR scan in  CCA

**Comment by andreievg** [comment]
> I'm keen to get into 1.5.00 just because it's taken so long to get the release out and I'm concerned that if I don't get this in now, it'll be ages before we can implement the QR scan in CCA

Oki. I'll re-test when i back in the office, thank you

File: client/packages/electron/src/electron.ts [hunk 4]
```diff
@@ -183,6 +184,11 @@ const tryToConnectToServer = (window: BrowserWindow, server: FrontEndHost) => {
 };
 
 const connectToServer = (window: BrowserWindow, server: FrontEndHost) => {
+  // translate loopback addresses to allow for clients, such as CCA to connect
+  // to the API by IP address
+  if (server.isLocal && isLoopback(server.ip)) {
```
Comment (mark-prins) [inline client/packages/electron/src/electron.ts:189]: this is the equivalent fix for electron

File: client/packages/electron/src/electron.ts [hunk 5]
```diff
@@ -446,6 +456,40 @@ const helpMenu: MenuItemConstructorOptions = {
         );
       },
     },
+    {
+      label: 'Clear Data',
+      click: () => {
```
Comment (mark-prins) [inline client/packages/electron/src/electron.ts:461]: added the dev menu items as we've had some issues raised with desktop and it's impossible to troubleshoot. May remove again once we've got to the bottom of things.

**Review by andreievg — APPROVED** [review]
Thanks for changes @mark-prins, looks good to me. Tested with electorn and android, works nicely.

**Comment by adamdewey** [comment]
Just wondering, should we perhaps include the `https://` in the QR return? Will there ever be a case where CCA needs to distinguish between http and https?

telegram-cloud-photo-size-1-5107095614415547639-y

**Comment by adamdewey** [comment]
Oh wait, maybe it does already and it's just the way my phone is displaying it?

**Comment by adamdewey** [comment]
Oh yeah, scratch that - it's just my phone being awkward

**Pull request merged**